server/src/internalClusterTest/java/org/elasticsearch/health/GetHealthCancellationIT.java

```diff
@@ -0,0 +1,157 @@
+import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.transport.netty4.Netty4Plugin;
+
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.action.support.ActionTestUtils.wrapAsRestResponseListener;
+import static org.elasticsearch.test.TaskAssertions.assertAllCancellableTasksAreCancelled;
+import static org.elasticsearch.test.TaskAssertions.assertAllTasksHaveFinished;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, numClientNodes = 0)
+public class GetHealthCancellationIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(getTestTransportPlugin(), MockTransportService.TestPlugin.class);
+    }
+
+    @Override
+    protected Settings nodeSettings(int ordinal, Settings otherSettings) {
+        return Settings.builder()
+            .put(super.nodeSettings(ordinal, otherSettings))
+            .put(NetworkModule.HTTP_DEFAULT_TYPE_SETTING.getKey(), Netty4Plugin.NETTY_HTTP_TRANSPORT_NAME)
+            .build();
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false; // enable http
+    }
+
+    public void testCancellation() throws Exception {
+        internalCluster().startMasterOnlyNode(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "master_node").build());
+        internalCluster().startDataOnlyNode(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "data_node").build());
+
+        final CountDownLatch tasksBlockedLatch = new CountDownLatch(1);
+        final SubscribableListener<Void> fetchHealthInfoRequestReleaseListener = new SubscribableListener<>();
+        for (TransportService transportService : internalCluster().getInstances(TransportService.class)) {
+            ((MockTransportService) transportService).addRequestHandlingBehavior(
+                FetchHealthInfoCacheAction.NAME,
+                (handler, request, channel, task) -> {
+                    tasksBlockedLatch.countDown();
+                    fetchHealthInfoRequestReleaseListener.addListener(
+                        ActionListener.wrap(ignored -> handler.messageReceived(request, channel, task), e -> {
+                            throw new AssertionError("unexpected", e);
+                        })
+                    );
+                }
+            );
+        }
+
+        final ClusterService clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final CountDownLatch healthNodeFoundLatch = new CountDownLatch(1);
+        AtomicReference<DiscoveryNode> healthNodeReference = new AtomicReference<>();
```

Comment:
Pretty sure you're just simulating a `PlainActionFuture` here.

Reply:
++